firedrake/mg/mesh.py

```diff
@@ -201,6 +201,36 @@ def ExtrudedMeshHierarchy(base_hierarchy, layers, kernel=None, layer_height=None
                          refinements_per_level=base_hierarchy.refinements_per_level,
                          nested=base_hierarchy.nested)
 
+def ExtrudedMeshHierarchyUniform(base_hierarchy, base_layers, height, kernel=None,
+                          extrusion_type='uniform', gdim=None,
+                          mesh_builder=firedrake.ExtrudedMesh):
+    """Build a hierarchy of extruded meshes by extruding a hierarchy of meshes.
+
+    :arg base_hierarchy: the unextruded base mesh hierarchy to extrude.
+    :arg mesh_builder: function used to turn a :class:`~.Mesh` into an
+       extruded mesh. Used by pyadjoint.
+
+    See :func:`~.ExtrudedMesh` for the meaning of the remaining parameters.
```

Comment:
This docstring doesn't really tell me the difference between this version and just an `ExtrudedMeshHierarchy`. How about, if we unify (as suggested by @dham) the interface:

```python
def ExtrudedMeshHierarchy(base_hierarchy, height, *layers, *, extrusion_type='uniform', gdim=None, mesh_builder=firedrake.ExtrudedMesh):
   """Build a hierarchy of extruded meshes by extruding a hierarchy of meshes.

   :arg base_hierarchy: The unextruded base mesh hierarchy to extrude.
   :arg height: Height of the extruded mesh.
   :arg layers: number of layers for each extruded mesh. Either pass a single integer (for same number of layers on every level), or else as many as there are meshes in the base hierarchy.

    .. note::

       If different numbers of layers are selected on each level, each coarse mesh must have the same number (or fewer levels) as its fine mesh counterpart, and the number of layers on successive levels must be integer-divisible.
"""
    try:
        layers, = layers
        layers = itertools.repeat(layers)
    except ValueError:
        if len(layers) != len(base_hierarchy):
            raise ValueError("Must either provide single layer for all meshes or layers for each mesh")
        if any(c >= f for c, f in zip(layers, layers[1:])):
            raise ValueError("Coarser meshes cannot have more levels")
        if any(f / c != f // c for c, f in zip(layers, layers[1:])):
            raise ValueError("Refinement ratio in extruded direction must be an integer")
    meshes = [mesh_builder(mesh, layer, ...) for mesh, layer in zip(base_hierarchy._meshes, layers)]
```

Reply:
Although I note this does a really weird thing if `refinements_per_level` of the base hierarchy is not 1. I need to think about that.

Reply:
I don't like that one would have to do something like

`
base = UnitSquareMesh(10, 10, quadrilateral=True)
basemh = MeshHierarchy(base, 2)
mh = ExtrudedMeshHierarchy(basemh, height=1, layers=[10 * 2**i for i in range(2)])
`

It seems ugly to have to specify the number of refinements twice (once in the base hierarchy, and once in the list of layers). This is doubly so when the ratio has to be an integer anyway. I think we should just specify the ratio and compute the layers for each refinement in the base hierarchy.

Reply:
Well, the ratio need not be 2 (I think). If we are going to allow coarsening in the extruded direction, I think it makes sense to support not just uniform refinement (or no refinement). e.g. if you want to do Poisson on thin domains, it makes sense to go from N layers to 1 layer in one step, and then solve 2D problems.

Why not offer:

```
EMH(..., base_layer=N, refinement_ratio=P, layers=[list_of_layers])
```

And check that user either provides base_layer and refinement_ratio (the latter defaulting to 1 as presently?) or list_of_layers?

Reply:
The code now implements the API @wence- suggests.